query_execution.go

```diff
@@ -0,0 +1,889 @@
+			Locations: locs,
+			Extensions: map[string]interface{}{
+				"selectionSet": formatSelectionSetSingleLine(q.ctx, q.schema, step.SelectionSet),
+			},
+		})
+	}
+
+	return outputErrs
+}
+
+// The insertionPoint represents the level a piece of data should be inserted at, relative to the root of the root step's data.
+// However results from a boundary query only contain a portion of that tree. For example, you could
+// have insertionPoint: ["foo", "bar", "movies". "movie", "compTitles"], with the below example as the boundary result we're
+// crawling for ids:
+// [
+// 	 {
+//     "_id": "MOVIE1",
+//     "compTitles": [
+//       {
+//   	   "_id": "1"
+// 		 }
+//	   ]
+//   }
+// ]
+//
+// We therefore cannot use the insertionPoint as is in order to extract the boundary ids for the next child step.
+// This function trims the insertionPoint up until we find a key that exists in both the boundary result and insertionPoint.
+// When a match is found, the remainder of the insertionPoint is used, which in this case is only ["compTitles"].
+// This logic is only needed when we are already in a child step, which itself contains it's own child steps.
+func trimInsertionPointForNestedBoundaryStep(data []interface{}, childInsertionPoint []string) ([]string, error) {
+	if len(data) < 1 {
+		return nil, fmt.Errorf("no boundary results to process")
+	}
+
+	firstBoundaryResult, ok := data[0].(map[string]interface{})
+	if !ok {
+		return nil, fmt.Errorf("a single boundary result should be a map[string]interface{}")
+	}
+	for i, point := range childInsertionPoint {
+		_, ok := firstBoundaryResult[point]
+		if ok {
+			return childInsertionPoint[i:], nil
+		}
+	}
+	return nil, fmt.Errorf("could not find any insertion points inside boundary data")
+}
+
+func executeBrambleStep(queryPlanStep *QueryPlanStep) (*executionResult, error) {
+	result, err := buildTypenameResponseMap(queryPlanStep.SelectionSet, queryPlanStep.ParentType)
+	if err != nil {
+		return nil, err
+	}
+
+	return &executionResult{
+		ServiceURL:     internalServiceName,
+		InsertionPoint: []string{},
+		Data:           result,
+	}, nil
+}
+
+func buildTypenameResponseMap(selectionSet ast.SelectionSet, parentTypeName string) (map[string]interface{}, error) {
+	result := make(map[string]interface{})
+	for _, field := range selectionSetToFields(selectionSet) {
+		if field.SelectionSet != nil {
+			if field.Definition.Type.NamedType == "" {
+				return nil, fmt.Errorf("buildTypenameResponseMap: expected named type")
+			}
+
+			var err error
+			result[field.Alias], err = buildTypenameResponseMap(field.SelectionSet, field.Definition.Type.Name())
+			if err != nil {
+				return nil, err
+			}
+		} else {
+			if field.Name != "__typename" {
+				return nil, fmt.Errorf("buildTypenameResponseMap: expected __typename")
+			}
+			result[field.Alias] = parentTypeName
+		}
+	}
+	return result, nil
+}
+
+func fragmentImplementsAbstractType(schema *ast.Schema, objectType, interfaceType string) bool {
```

Comment:
@lucianjon `interfaceType` is a bit misleading here since it can also be a union